crates/wysiwyg/src/composer_model/hyperlinks.rs

```diff
@@ -48,6 +49,19 @@ where
         }
     }
 
+    pub fn set_link_suggestion(
+        &mut self,
+        link: S,
+        text: S,
+        suggestion: SuggestionPattern,
+    ) -> ComposerUpdate<S> {
+        self.do_replace_text_in(S::default(), suggestion.start, suggestion.end);
+        self.state.start = Location::from(suggestion.start);
+        self.state.end = self.state.start;
+        self.set_link_with_text(link, text);
```

Comment:
Do you think representing pills internally as a link node is enough, or should we consider something more flexible?

I'm asking because pills might need some custom behavior like removing the whole pill when backspacing from the end. So, would it be better to represent pills internally as an abstract 'pill' and then convert it to a link during HTML rendering?

Reply:
Yeah, we talked about this a bit. It is done that way to ease up parsing. If we e.g. parse HTML to edit an old message including pills, pills in that HTML are just links, and if we want the Rust library to be aware this is represented as a pill, we'll have to add some kind of custom HTML attributes manually before parsing for the library to understand these are not standard links. 
This would mean we'd have to both run some permalink detection before sending the HTML to Rust, and run permalink detection again when Rust sends back the current HTML. And AFAIK it seems that deleting the entire pill with backspace is almost the only custom behaviour required. We are for now trying to implement this behaviour in the platform specific code (on iOS it just works and remove the entire pill already, because we do some selection if needed before backspace in case of multiple code points emojis - in the case of pills, they are a 1-length item in the attributed string, and we will just calculate the real length in HTML and remove that, without any change to the current backspace implementation)

The main idea is to try out without Rust having that knowledge, because Rust shouldn't know if some platform wants to represent these links in some other way.
We might move to Rust having that knowledge if it causes too much problems on a specific platform.

Reply:
Ok, I see why we'd represent the pill as a link internally :+1: 

About the backspace handling though, is it right that this is the app's responsibility? Let's say an app does not implement it and uses the default behaviour. Am I right that backspacing will result in a misleading link? For example

**Before**
```
<a href="matrix.to/alice">@alice</a>
```
**After two backspaces**
```
<a href="matrix.to/alice">@ali</a> // misleading!
```

Reply:
Fair point. Well, hopefully most clients will implement it.

In the case where these are represented as standard links, isn't it weird from a composing perspective that a standard link text component can be edited freely and what was supposed to be a mention but is represented in the exact same fashion can't ? And what would happen if I move the cursor inside the text of that mention link ? Do we just don't allow it from the Rust side ? 

Also, same thing could actually happen if the permalink detection doesn't work for some reason (and we can't prevent it in that case).

The other important thing is that, in Matrix, mentions are links, it's not just an internal representation of a pill. I would rather say that Pills are an external/UI representation of a link.
I think we can still see the same kind of issue if we have a Rust internal type for Pills. If a client doesn't care about pill representation, it's fair to assume that it won't provide Rust with permalink detection for parsing either, so anytime we parse the HTML (e.g. editing a message), the Rust internal type for Pills will not be here anyway, and we'll end up with just a standard link.

Reply:
I can see the argument that the pill representation is a UI/app concern given that mentions are just links according to the spec.

It could still be useful to have the expected behaviour be the default IMO. But given the edge cases you mentioned, I'd agree we can keep it simple to start and add it later.